facebook/delphiFacebook/R/contingency_aggregate.R

```diff
@@ -0,0 +1,361 @@
+    
+    # Add implied geo_level to each group_by. Order alphabetically. Replace 
+    # geo_level with generic "geo_id" var. Remove duplicate grouping vars.
+    if (length(geo_level) > 1) {
+      stop('more than one geo type provided for a single aggregation')
+      
+    } else if (length(geo_level) == 0) {
+      # Presume national grouping
+      geo_level <- "nation"
+      aggregations$group_by[agg_ind][[1]] <- 
+        sort(append(aggregations$group_by[agg_ind][[1]], "geo_id"))
+      
+    } else {
+      aggregations$group_by[agg_ind][[1]][
+        aggregations$group_by[agg_ind][[1]] == geo_level] <- "geo_id"
+      aggregations$group_by[agg_ind][[1]] <- 
+        sort(unique(aggregations$group_by[agg_ind][[1]]))
+    }
+    
+    aggregations$geo_level[agg_ind] <- geo_level
+    
+    # Multiple choice metrics should also be included in the group_by vars
+    if (startsWith(aggregations$metric[agg_ind], "mc_")) {
+      if ( !(aggregations$metric[agg_ind] %in% 
+             aggregations$group_by[agg_ind][[1]]) ) {
+        aggregations$group_by[agg_ind][[1]] <- 
+          c(aggregations$group_by[agg_ind][[1]], aggregations$metric[agg_ind])
+      }
+    }
+  }
+  
+  # Convert most columns being used in aggregations to the appropriate format.
+  # Multiple choice and multi-select used for grouping are left as-is.
+  #### TODO: multi-select responses should be converted into set of binary cols 
+  #### if being used for grouping; low priority since using a multi-select as a
+  #### grouping variable should be rare.
+  agg_groups <- unique(aggregations$group_by)
+  group_cols_to_convert <- unique(do.call(c, agg_groups))
+  group_cols_to_convert <- group_cols_to_convert[startsWith(group_cols_to_convert, "b_")]
+  
+  metric_cols_to_convert <- unique(aggregations$metric)
+  
+  #### TODO: move to separate function. Remove or condense all the `output[[x]]`; messy
+  for (col_var in c(group_cols_to_convert, metric_cols_to_convert)) {
+    if ( is.null(df[[col_var]]) ) {
+      # Column not defined. Remove all aggregations that use it as metric or in
+      # group_by variables
+      aggregations <- aggregations[aggregations$metric != col_var & 
+                                     !mapply(aggregations$group_by, 
+                                             FUN=function(x) {col_var %in% x}), ]
+      next
+    }
+    
+    if (startsWith(col_var, "b_")) { # Binary
+      output <- code_binary(df, aggregations, col_var)
+      df <- output[[1]]
+      aggregations <- output[[2]]
+      
+    } else if (startsWith(col_var, "ms_")) { # Multiselect
+      output <- code_multiselect(df, aggregations, col_var)
+      df <- output[[1]]
+      aggregations <- output[[2]]
+      
+    } else if (startsWith(col_var, "n_")) { # Numeric free response
+      output <- code_numeric_freeresponse(df, aggregations, col_var)
+      df <- output[[1]]
+      aggregations <- output[[2]]
+    }
```

Comment:
should there be an "else throw an error for bad column name" case here?

Reply:
We shouldn't need an error here; there's an if block just before this that skips this section if the column doesn't exist.